go/vt/wrangler/materializer.go

```diff
@@ -941,22 +941,50 @@ func (wr *Wrangler) buildMaterializer(ctx context.Context, ms *vtctldatapb.Mater
 			}
 		}
 	}
-
+	isPartial := false
 	sourceShards, err := wr.sourceTs.GetServingShards(ctx, ms.SourceKeyspace)
 	if err != nil {
 		return nil, err
 	}
+	if len(ms.SourceShards) > 0 {
+		isPartial = true
+		var sourceShards2 []*topo.ShardInfo
+		for _, shard := range sourceShards {
+			for _, shard2 := range ms.SourceShards {
+				if shard.ShardName() == shard2 {
+					sourceShards2 = append(sourceShards2, shard)
+					break
+				}
+			}
+		}
+		sourceShards = sourceShards2
+	}
 	targetShards, err := wr.ts.GetServingShards(ctx, ms.TargetKeyspace)
 	if err != nil {
 		return nil, err
 	}
-
+	if len(ms.SourceShards) > 0 {
+		var targetShards2 []*topo.ShardInfo
+		for _, shard := range targetShards {
+			for _, shard2 := range ms.SourceShards {
+				if shard.ShardName() == shard2 {
+					targetShards2 = append(targetShards2, shard)
+					break
+				}
+			}
+		}
+		targetShards = targetShards2
+	}
+	if len(sourceShards) == 0 || len(targetShards) == 0 {
+		return nil, fmt.Errorf("no source and/or target shards specified for workflow %s ", ms.Workflow)
+	}
```

Comment:
Ah, this answers my previous two questions :). IMO it would be better to error for each one as we return sooner (do less work) and we can provide a more specific error.